pkg/builder/controller.go

```diff
@@ -54,55 +59,246 @@ const (
 
 // Builder builds a Controller.
 type Builder struct {
-	forInput         ForInput
-	ownsInput        []OwnsInput
-	watchesInput     []WatchesInput
+	forInput         forInput
+	forInputErr      error
+	ownsInput        []watcher
+	watchesInput     []watcher
 	mgr              manager.Manager
-	globalPredicates []predicate.Predicate
+	globalPredicates []predicate.Predicate[client.Object]
 	ctrl             controller.Controller
 	ctrlOptions      controller.Options
 	name             string
 }
 
+func newObjectPredicate[T client.ObjectConstraint](p predicate.Predicate[T]) predicate.Predicate[client.Object] {
```

Comment:
This is heavily work in progress, but I wanted to push changes to make sure this approach is okay
Since controller is able to watch on multiple resources, the only way I could think of (since we can't use generic methods) is to create a generic function that will convert this generic type to a `client.Object`
That way, the complexity is internal, and we use `client.Object` interface internally. The caller still can leverage generics to use those types

The only thing is that the project needs to be go version 1.20, because they added interfaces to be comparable in that version

Reply:
This does not work now, there are many tests where I'd have to provide types and similar
The controller watch is also something that needs to be changed or provided another method like `controller.WatchType(WithWatch()` that can be a generic function returning wrapped objects

Reply:
Since many types can handle multiple types, wrapping them in interfaces was the only thing that I could think of to provide a nice API... Some types can be improved, but the only thing I wanted to ask is, is there a different approach I should take?

Reply:
This is the issue I was referring to in https://github.com/kubernetes-sigs/controller-runtime/pull/2265#discussion_r1162854237. We have to change how this works in order to accomodate generics. The source needs to get the handler and predicates at its construction time rather than at `Start` time, because using `Start` entails passing it through the `controller` and that would entail that we can only support one type per controller.

Reply:
I see, I'll work on it